core/node/l1_recovery/src/l1_fetcher/l1_fetcher.rs

```diff
@@ -0,0 +1,1083 @@
+                        }
+                        _ => {
+                            tracing::error!("Failed to parse calldata: {e}, encountered on block: {block_number}");
+                        }
+                    }
+                    return Err(e);
+                }
+            }
+        }
+    }
+
+    /// Get a specified transaction on L1 by its hash.
+    async fn get_transaction_by_hash(
+        eth_client: &Box<DynClient<L1>>,
+        hash: H256,
+    ) -> Result<Transaction> {
+        match L1Fetcher::retry_call(
+            || L1Fetcher::query_client(eth_client).get_tx(hash),
+            L1FetchError::GetTx,
+        )
+        .await
+        {
+            Ok(Some(tx)) => Ok(tx),
+            Ok(None) => Err(anyhow!("unable to find transaction with hash: {}", hash)),
+            Err(e) => Err(e),
+        }
+    }
+
+    fn query_client(eth_client: &Box<DynClient<L1>>) -> &dyn EthInterface {
+        eth_client
+    }
+    /// Get the last published L1 block.
+    async fn get_last_l1_block_number(eth_client: &Box<DynClient<L1>>) -> Result<U64> {
+        let last_block = L1Fetcher::retry_call(
+            || L1Fetcher::query_client(eth_client).block(BlockId::Number(BlockNumber::Finalized)),
+            L1FetchError::GetEndBlockNumber,
+        )
+        .await?;
+
+        last_block
+            .expect("last block must be present")
+            .number
+            .ok_or_else(|| anyhow!("found latest block, but it contained no block number"))
+    }
+
+    async fn retry_call<T, E, Fut>(callback: impl Fn() -> Fut, err: L1FetchError) -> Result<T>
+    where
+        Fut: Future<Output = Result<T, E>>,
+        E: std::fmt::Display,
+    {
+        sleep(Duration::from_millis(100)).await;
+        for attempt in 1..=100 {
+            match callback().await {
+                Ok(x) => return Ok(x),
+                Err(e) => {
+                    tracing::error!("attempt {attempt}: failed to fetch from L1: {e}");
+                    sleep(Duration::from_millis(
+                        attempt * 1000 + random::<u64>() % 500,
+                    ))
+                    .await;
+                }
+            }
+        }
+        Err(err.into())
+    }
+}
+
+pub async fn parse_last_committed_l1_batch(
+    commit_candidates: &[Function],
+    calldata: &[u8],
+) -> Result<StoredBatchInfo, ParseError> {
+    let commit_fn = commit_candidates
+        .iter()
+        .find(|f| f.short_signature() == calldata[..4])
+        .unwrap();
+    let mut parsed_input = commit_fn.decode_input(&calldata[4..]).unwrap();
+
+    let _new_blocks_data = parsed_input.pop().unwrap();
+    let stored_block_info = parsed_input.pop().unwrap();
+    return Ok(StoredBatchInfo::from_token(stored_block_info).unwrap());
+}
+
+pub async fn parse_calldata(
+    protocol_versioning: &ProtocolVersioning,
+    l1_block_number: u64,
+    commit_candidates: &[Function],
+    calldata: &[u8],
```

Comment:
this parsing looks very manual. Can't we make it more robust by using ethabi library somehow? Also l1_contracts_interface could be a better place for this logic.